### PR TITLE
make hyper tuples canonical

### PR DESCRIPTION
**Pull request by smichr**
#### Brief description of what is fixed or changed

The hyper and meijerg tuples are used as elements in products and can be made canonical in order. Also, simplification rules for hyper allow cancellation of common elements in numerator and denominator.

see [basic properties](https://en.wikipedia.org/wiki/Generalized_hypergeometric_function)

#### Release Notes

* functions
    * hyper and meijerg function now return canonically ordered tuples; hyper will automatically remove redundant args

### CODE REVIEW

File: sympy/functions/special/hyper.py [hunk 1]
```diff
@@ -191,6 +195,22 @@ class hyper(TupleParametersBase):
 
     def __new__(cls, ap, bq, z, **kwargs):
         # TODO should we check convergence conditions?
+        ca = Counter(Tuple(*ap))
+        cb = Counter(Tuple(*bq))
+        for i in ca:
+            if i in cb and i.is_nonnegative:
+                if ca[i] > cb[i]:
```
Comment (oscarbenjamin) [inline sympy/functions/special/hyper.py:202]: Will this raise for symbolic expressions?

Comment (smichr) [reply sympy/functions/special/hyper.py:202]: I thought the same thing when reviewing, but no: a counter's values are ints and Counter is happy to collect on symbols:

```python
>>> Counter(var('x x y'))
Counter({x: 2, y: 1})
```

Comment (oscarbenjamin) [reply sympy/functions/special/hyper.py:202]: Is this like:
```python
In [64]: c1 = Counter([1, 2, 3, 1])

In [65]: c2 = Counter([1, 3, 3, 1])

In [66]: c1 - (c1 & c2)
Out[66]: Counter({2: 1})

In [67]: c2 - (c1 & c2)
Out[67]: Counter({3: 1})
```

Comment (smichr) [reply sympy/functions/special/hyper.py:202]: That's a nice feature, more set like than I knew. Code is being updated.

**Comment by smichr** [comment]
Should the elimination of common items in `hyper` be part of the simplify process or limited by the evaluate flag?

**Comment by oscarbenjamin** [comment]
> Should the elimination of common items in `hyper` be part of the simplify process or limited by the evaluate flag?

It should be limited by the eliminate flag. Otherwise I think a quick test based on `==` is okay for standard evaluation.

For explicit simplification there could be something based on `Eq` that might be more comprehensive at finding equivalent expressions to be cancelled.

**Comment by oscarbenjamin** [comment]
> > the eliminate flag
> 
> Are you recommending that I add this flag?

I meant evaluate!

File: sympy/functions/special/hyper.py [hunk 2]
```diff
@@ -120,17 +124,17 @@ class hyper(TupleParametersBase):
 
     >>> from sympy import hyper
     >>> from sympy.abc import x, n, a
-    >>> hyper((1, 2, 3), [3, 4], x)
-    hyper((1, 2, 3), (3, 4), x)
+    >>> h = hyper((1, 2, 3), [3, 4], x); h
+    hyper((1, 2), (4,), x)
```
Comment (smichr) [inline sympy/functions/special/hyper.py:129]:
```suggestion
    >>> h = hyper((1, 2, 3), [3, 4], x); h
    hyper((1, 2), (4,), x)
    >>> hyper((3, 1, 2), [3, 4], x, evaluate=False)  # don't remove duplicates
    hyper((1, 2, 3), (3, 4), x)
```

File: sympy/functions/special/hyper.py [hunk 3]
```diff
@@ -191,6 +198,18 @@ class hyper(TupleParametersBase):
 
     def __new__(cls, ap, bq, z, **kwargs):
         # TODO should we check convergence conditions?
+        if kwargs.pop('evaluate', global_parameters.evaluate)
```
Comment (smichr) [inline sympy/functions/special/hyper.py:201]:
```suggestion
        if kwargs.pop('evaluate', global_parameters.evaluate):
```

**Pull request merged**